File: Project1/Baitoan3/a/self_conjugate_partition_k_parts.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <cstddef>

namespace core {
    /**
     * @brief Generate all integer partitions of n into exactly k parts (non-increasing order)
     * @param n The integer to partition
     * @param k The number of parts
     * @param result The list to store resulting partitions
     * @param currentPartition The current partition (for recursion)
     * @param maxVal The maximum value allowed for the current part
     */
    void generatePartitionsWithKParts(int n, int k, std::vector<std::vector<int>>& result, std::vector<int>& currentPartition, int maxVal) {
        if (k == 0) {
            if (n == 0) {
                result.push_back(currentPartition);
            }
            return;
        }
        if (n <= 0 || k <= 0) return;
        if (n < k) return;
        const int upperBound = std::min(maxVal, n - (k - 1));
        for (int val = upperBound; val >= 1; --val) {
            currentPartition.push_back(val);
            generatePartitionsWithKParts(n - val, k - 1, result, currentPartition, val);
            currentPartition.pop_back();
        }
    }
    /**
     * @brief Check if a partition is self-conjugate
     * @param partition The partition (non-increasing order)
     * @return true if self-conjugate, false otherwise
     */
    bool isSelfConjugate(const std::vector<int>& partition) {
        if (partition.empty()) return true;
        const int maxVal = partition[0];
        std::vector<int> conjugateParts;
        for (int j = 1; j <= maxVal; ++j) {
            int count = 0;
            for (int partVal : partition) {
                if (partVal >= j) ++count;
                else break;
            }
            if (count > 0) conjugateParts.push_back(count);
            else break;
        }
        return partition == conjugateParts;
    }
} // namespace core

namespace utils {
    /**
     * @brief Print a partition in the format [a, b, c]
     * @param partition The partition to print
     */
    void printPartition(const std::vector<int>& partition) {
        std::cout << "[";
        for (std::size_t i = 0; i < partition.size(); ++i) {
            std::cout << partition[i];
            if (i + 1 < partition.size()) std::cout << ", ";
        }
        std::cout << "]\n";
    }
} // namespace utils

int main() {
    int n = 0;
    int k = 0;
    std::ifstream fin("input.txt");
    if (!fin) {
        std::cerr << "Error: Cannot open input.txt\n";
        return 1;
    }
    fin >> n >> k;
    fin.close();
    if (n < 0 || k < 0) {
        std::cout << "n and k must be non-negative integers.\n";
        return 0;
    }
    if (k > n && n > 0) {
        std::cout << "No partition of " << n << " into " << k << " parts.\n";
        return 0;
    }
    if (k == 0 && n > 0) {
        std::cout << "No partition of " << n << " into " << k << " parts.\n";
        return 0;
    }
    if (k == 0 && n == 0) {
        std::cout << "Number of self-conjugate partitions of 0 into 0 parts: 1\n";
        std::cout << "Partitions: [[]]\n";
        return 0;
    }
    std::vector<std::vector<int>> allKPartitions;
    std::vector<int> currentPartition;
    core::generatePartitionsWithKParts(n, k, allKPartitions, currentPartition, n);
    std::vector<std::vector<int>> selfConjugateKPartitions;
    for (const auto& partition : allKPartitions) {
        if (core::isSelfConjugate(partition)) {
            selfConjugateKPartitions.push_back(partition);
        }
    }
    std::cout << "Number of self-conjugate partitions of " << n << " into " << k << " parts: " << selfConjugateKPartitions.size() << "\n";
    std::cout << "Partitions:\n";
    if (selfConjugateKPartitions.empty()) {
        std::cout << "None.\n";
    } else {
        for (const auto& partition : selfConjugateKPartitions) {
            utils::printPartition(partition);
        }
    }
    return 0;
} 

File: Project1/Baitoan3/a/self_conjugate_partition_k_parts.py
import sys

def generate_partitions_k_parts(n, k):
    partitions = []
    current_partition = []
    def _generate(remaining_sum, num_parts_left, max_val_for_this_part):
        if num_parts_left == 0:
            if remaining_sum == 0:
                partitions.append(list(current_partition))
            return
        if remaining_sum <= 0 or num_parts_left <= 0:
            return
        if remaining_sum < num_parts_left:
            return
        upper_bound = min(max_val_for_this_part, remaining_sum - (num_parts_left - 1))
        for val in range(upper_bound, 0, -1):
            current_partition.append(val)
            _generate(remaining_sum - val, num_parts_left - 1, val)
            current_partition.pop()
    _generate(n, k, n)
    return partitions

def is_self_conjugate(partition):
    if not partition:
        return True
    max_val = partition[0]
    conjugate_parts = []
    for j in range(1, max_val + 1):
        count = 0
        for part_val in partition:
            if part_val >= j:
                count += 1
            else:
                break
        if count > 0:
            conjugate_parts.append(count)
        else:
            break
    return partition == conjugate_parts

def main():
    # Read input from input.txt
    try:
        with open("input.txt", "r", encoding="utf-8") as f:
            lines = f.read().split()
            n = int(lines[0])
            k = int(lines[1])
    except Exception as e:
        print("Error reading input.txt:", e)
        sys.exit(1)
    if n < 0 or k < 0:
        print("n and k must be non-negative integers.")
        return
    if k > n and n > 0:
        print(f"No partition of {n} into {k} parts.")
        return
    if k == 0 and n > 0:
        print(f"No partition of {n} into {k} parts.")
        return
    if k == 0 and n == 0:
        print("Number of self-conjugate partitions of 0 into 0 parts: 1")
        print("Partitions: [[]]")
        return
    all_k_partitions = generate_partitions_k_parts(n, k)
    self_conjugate_k_partitions = []
    for p in all_k_partitions:
        if is_self_conjugate(p):
            self_conjugate_k_partitions.append(p)
    print(f"Number of self-conjugate partitions of {n} into {k} parts: {len(self_conjugate_k_partitions)}")
    print("Partitions:")
    if not self_conjugate_k_partitions:
        print("None.")
    else:
        for p in self_conjugate_k_partitions:
            print(p)

if __name__ == "__main__":
    main() 